### PR TITLE
Ecqm QDM models

### PR DESCRIPTION
**Pull request by kchapple**
Adds QDM Model Builder and a set of services to query OpenEMR DB and generate QDM Models to serialize and pass to calculator. 

The file tests/Tests/ECQM/MeasureResultsTest.php has the example code of how to build a result for a measure for one patient. Another example exists in the cqm tools AdminController.php.

Adds some testing to PHPUnit to attempt to verify our results match results in Cypress demo, but this proved troublesome because I don't have the effective dates. Models look pretty good and complete to satisfy the initial set of measures:
CMS69 - BMI screen and f/u
CMS122 - DM A1C poor control
CMS124 - cervical ca screen
CMS125 - breast ca screen
CMS127 - pneumo vaccine
CMS130 - colon ca screen
CMS138 - tobacco use screen and cessation intervention
CMS147 - flu vaccine
CMS165 - HTN control

### CODE REVIEW

**Comment by sjpadgett** [comment]
@mdsupport Can you go ahead and update this PR?

**Comment by mdsupport** [comment]
@sjpadgett Can you check if the PR I sent is what you wanted?

**Comment by sjpadgett** [comment]
@mdsupport That should take care of php stylings but the big concern is this PR is out of date with master with conflicts.
Most likely the package managers and globals.

**Comment by adunsulag** [comment]
@sjpadgett I think you were meaning to tag @kchapple originally, not @mdsupport.

**Comment by sjpadgett** [comment]
LOL! You see this is why I shouldn't be allowed on computers anymore!
@mdsupport So sorry and thank you for pitching in anyway. Good guy!
@kchapple Let's try this again.

**Comment by sjpadgett** [comment]
@kchapple What version PHP are you testing with? Note on PHP 8.1 Generate Models
```
[24-Jan-2022 15:13:21 America/New_York] PHP Fatal error:  Uncaught TypeError: Cannot assign null to property Laminas\Code\Generator\DocBlockGenerator::$shortDescription of type string in C:\xampp\htdocs\openemr_mi2\vendor\laminas\laminas-code\src\Generator\DocBlockGenerator.php:122
Stack trace:
#0 C:\xampp\htdocs\openemr_mi2\vendor\laminas\laminas-code\src\Generator\DocBlockGenerator.php(72): Laminas\Code\Generator\DocBlockGenerator->setShortDescription(NULL)
#1 C:\xampp\htdocs\openemr_mi2\src\Cqm\Generator.php(223): Laminas\Code\Generator\DocBlockGenerator::fromArray(Array)
#2 C:\xampp\htdocs\openemr_mi2\interface\modules\custom_modules\oe-module-cqm\src\Controllers\AdminController.php(69): OpenEMR\Cqm\Generator->execute()
#3 C:\xampp\htdocs\openemr_mi2\vendor\mi-squared\mi-framework\src\Mi2\Framework\App.php(89): Mi2\Cqm\Controllers\AdminController->_action_generate_models()
#4 C:\xampp\htdocs\openemr_mi2\vendor\mi-squared\mi-framework\src\Mi2\Framework\App.php(60): Mi2\Framework\App->perform(Object(Mi2\Cqm\Controllers\AdminController), 'generate_models')
#5 C:\xampp\htdocs\openemr_mi2\interface\modules\custom_modules\oe-module-cqm\index.php(7): Mi2\Framework\App->run()
#6 {main}
  thrown in C:\xampp\htdocs\openemr_mi2\vendor\laminas\laminas-code\src\Generator\DocBlockGenerator.php on line 122
```

And Generate Json: looks like the property is `_openEmrPid` doesn't exist in Patient

```
[24-Jan-2022 15:17:13 America/New_York] PHP Fatal error:  Uncaught Exception: Property  does not exist on OpenEMR\Cqm\Qdm\Patient in C:\xampp\htdocs\openemr_mi2\src\Cqm\Qdm\BaseTypes\AbstractType.php:22
Stack trace:
#0 C:\xampp\htdocs\openemr_mi2\src\Services\Qdm\PatientService.php(21): OpenEMR\Cqm\Qdm\BaseTypes\AbstractType->__construct(Array)
#1 C:\xampp\htdocs\openemr_mi2\interface\modules\custom_modules\oe-module-cqm\src\Controllers\AdminController.php(78): OpenEMR\Services\Qdm\PatientService->makePatient('500')
#2 C:\xampp\htdocs\openemr_mi2\vendor\mi-squared\mi-framework\src\Mi2\Framework\App.php(89): Mi2\Cqm\Controllers\AdminController->_action_generate_patient()
#3 C:\xampp\htdocs\openemr_mi2\vendor\mi-squared\mi-framework\src\Mi2\Framework\App.php(60): Mi2\Framework\App->perform(Object(Mi2\Cqm\Controllers\AdminController), 'generate_patien...')
#4 C:\xampp\htdocs\openemr_mi2\interface\modules\custom_modules\oe-module-cqm\index.php(7): Mi2\Framework\App->run()
#5 {main}
  thrown in C:\xampp\htdocs\openemr_mi2\src\Cqm\Qdm\BaseTypes\AbstractType.php on line 22

```

File: src/Cqm/CqmClient.php [hunk 1]
```diff
@@ -25,20 +25,23 @@ class CqmClient extends HttpClient
...
-        $node = "node"; //$GLOBALS['node_binary'];
+        $node = $GLOBALS['node_binary'];
         $cmd = $this->servicePath;
 
         if (IS_WINDOWS) {
             $cmd = "start /B $node $cmd";
             return $cmd;
         } else {
+            $command = $node;
             $system = new System();
```
Comment (sjpadgett) [inline src/Cqm/CqmClient.php:36]: Need to test if the global is populated. Most admins are not going to know where the node binary is located much less know to populate the global. Installing node via NVS will not set an OS global path to binary like install via PPA or MSI on windows. This is generally how node is installed and should not require us setting a path to node binary. Even for MacOS. So at minimum we should set the node binary if missing in code e.g:

```
if (IS_WINDOWS) {
            if (empty($node)) {
                $node = 'node';
            }
            $cmd = "start /B $node $cmd";
            return $cmd;
        } else {
            if (empty($node)) {
                $node = 'nodejs';
            } 
... 
```

File: src/Cqm/Qdm/Traits/PatientExtension.php [hunk 2]
```diff
@@ -5,7 +5,7 @@
 trait PatientExtension
 {
     public $_fullName;
-    public $_openEmrPid;
+    public $_id;
```
Comment (sjpadgett) [inline src/Cqm/Qdm/Traits/PatientExtension.php:9]: Was this meant to change

Comment (kchapple) [reply src/Cqm/Qdm/Traits/PatientExtension.php:9]: This was me trying to preserve the PID through the CQM calculator to try to map results to openemr patients, but all extra fields are stripped.

File: sql/6_0_0-to-6_1_0_upgrade.sql [hunk 3]
```diff
@@ -1275,7 +1275,7 @@ ALTER TABLE `form_observation` ADD `ob_reason_status` VARCHAR(32) NULL;
 ALTER TABLE `form_observation` ADD `ob_reason_code` VARCHAR(32) NULL;
 #EndIf
 #IfMissingColumn form_observation ob_reason_text
-ALTER TABLE `form_observation` ADD `ob_reason_text` TEXT;
+ALTER TABLE `form_observation` ADD `ob_reason_text` TEXT NULL;
```
Comment (sjpadgett) [inline sql/6_0_0-to-6_1_0_upgrade.sql:1278]: this is conflict and can be removed from PR as current master is correct. Same for database below.

File: composer.json [hunk 4]
```diff
@@ -67,7 +67,8 @@
         "twig/twig": "3.3.4",
         "vlucas/phpdotenv": "5.4.0",
         "waryway/php-traits-library": "1.0.4",
-        "yubico/u2flib-server": "1.0.2"
+        "yubico/u2flib-server": "1.0.2",
+        "openemr/oe-module-cqm": "dev-main"
```
Comment (sjpadgett) [inline composer.json:71]: Not sure we want the tool in production version. Really should only be installed when needed by admins for reporting period updates on new releases.

File: package.json [hunk 5]
```diff
@@ -97,7 +97,9 @@
     "select2": "4.0.13",
     "sortablejs": "1.14.0",
     "underscore": "1.13.1",
-    "validate.js": "0.13.1"
+    "validate.js": "0.13.1",
+    "oe-cqm-parsers": "github:openemr/oe-cqm-parsers",
+    "oe-cqm-service": "github:openemr/oe-cqm-service"
```
Comment (sjpadgett) [inline package.json:102]: This probably should be a dev dependency only. Again we don't need service in production.

**Comment by sjpadgett** [comment]
@kchapple In the past my experiences with submodules has been sometimes frustrating. It's been a couple years and appears git has done considerable work on the feature with this implementation seemly working okay.
Still, I don't think admins want to have to manage them. If this particular submodule was needed for the production environment, I'd might be willing to open a discussion because I can see where they'd be very helpful if we wanted to include custom modules with releases.

Still though, I don't believe we want to auto install oe-module-cqm as it will only be needed for release builds if new patient models are needed for new reporting periods.

So currently the only dependency packages we want in production would be those required to run the calculator or to support reporting.
Also, your mi2 mini framework used in the tool service needs to be either made part of openemr project or written out of service per Brady.

Recommend fixing the existing conflicts so we can get this part of project in core where I can help with debugging and integration.

**Comment by kchapple** [comment]
No prob, I’ll just remove it. I think it was just in the json file because I had it in there, I’ll have to ‘composer remove’ it before I check in, or just those files.

File: package.json (hunk 5 above)
Comment (kchapple) [reply package.json:102]: @sjpadgett This will be required to run reports. It's the eCQM service. Do you just mean the oe-cqm-parsers ?

Comment (sjpadgett) [reply package.json:102]: I unsure now. Just leave and I'll figure it out once I get to setting up for testing.

**Pull request closed without merging**